src/components/Input/Input.tsx

```diff
@@ -104,120 +110,123 @@ class Input extends AutoControlledComponent<Extendable<IInputProps>, any> {
      */
     renderInput: PropTypes.func,
 
+    /**
+     * A custom render function the wrapper slot.
+     *
+     * @param { React.ReactType } Component - The computed component for this slot.
+     * @param { object } props - The computed props for this slot.
+     * @param { ReactNode | ReactNodeArray } children - The computed children for this slot.
+     */
+    renderWrapper: PropTypes.func,
+
     /** Additional CSS styles to apply to the component instance.  */
     styles: PropTypes.oneOfType([PropTypes.object, PropTypes.func]),
 
+    /** The HTML input type. */
+    type: PropTypes.string,
+
     /** The value of the input. */
     value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
 
     /** Override for theme site variables to allow modifications of component styling via themes. */
     variables: PropTypes.oneOfType([PropTypes.object, PropTypes.func]),
+
+    /** Shorthand for the wrapper component */
+    wrapper: customPropTypes.itemShorthand,
   }
 
   static defaultProps = {
     as: 'div',
     type: 'text',
+    wrapper: 'div',
   }
 
   static autoControlledProps = ['value']
 
-  inputRef: any
+  state = { value: this.props.value || this.props.defaultValue || '' }
+
+  renderComponent({
+    ElementType,
+    classes,
+    rest: restProps,
+    styles,
+    variables,
+  }: IRenderResultConfig<IInputProps>) {
+    const { className, input, renderIcon, renderInput, renderWrapper, type, wrapper } = this.props
+    const { value = '' } = this.state
+    const [htmlInputProps, rest] = partitionHTMLProps(restProps)
+
+    return Slot.create(wrapper, {
+      defaultProps: {
+        as: ElementType,
+        className: cx(Input.className, className),
+        styles: styles.root,
+        ...rest,
+      },
+      overrideProps: {
+        children: (
```

Comment:
Why are we always overriding the wrapper's children? How will we allow the user to add custom children if they want? Or we don't support children API here?

Reply:
for an input the children API does not make sense;
we can definitely pass children as defaultProps but then **we will open up for the users scenarios where they can do:**

```jsx
<Input placeholder="Search..." />
<Input placeholder="Search..." wrapper={{ children: <Text content='random text instead of input' /> }} />
<Input placeholder="Search...">
  <input placeholder="Override haha" />
</Input>
<Input placeholder="Search...">
  <span>random span instead of input</span>
</Input>
```
**that would render:**
![screen shot 2018-10-17 at 15 05 52](https://user-images.githubusercontent.com/5442794/47088237-45ff2100-d21e-11e8-8233-9a1213195de5.png)

**by using `overrideProps` to pass `children` we do not allow the weird scenarios above and for the same inputs the rendered elements are:**
![screen shot 2018-10-17 at 15 07 24](https://user-images.githubusercontent.com/5442794/47088331-82328180-d21e-11e8-9890-9b597b4f15ee.png)

Reply:
discussed with @kuzhelov and came to the point that we should keep the consistency of the API and let the user pass `children` if they want; so I'm moving children to `defaultProps`;

no need to handle this with case `childrenExist` utility in the code as it's a side effect of the `Slot.create` method